doc/dev/crimson/poseidonstore.rst

```diff
@@ -0,0 +1,401 @@
+
+- Flush
+  - Flush WAL entries whenever needed
+
+
+Crash consistency
+------------------
+Large case
+- 1. Crash occurs right after writing Data blocks
+  - Data partition --> | Data blocks |
+  - We don't need to care this case. Data is not alloacted yet in reality. The blocks will be reused.
+- 2. Crash occurs right after WAL using atomic write command
+  - WAL --> | TxBegin A | Log Entry| TxEnd A |
+  - Data partition --> | Data blocks |
+  - Write procedure is completed, so there is no data loss or inconsistent state
+
+Small case
+- 1. Crash occurs right after writing WAL
+  - WAL --> | TxBegin A | Log Entry| TxEnd A |
+  - All data has been written
+
+
+Comparison
+----------
+- Best case (pre-allocation)
+  - Only need two writes on both WAL and Data partition without updating Object meta and Data bitmap
+- Worst case (flush happens)
+  - Without embedding, at least three writes are required additionally on Object meta, Data bitmap, and Data blocks
+- WAL needs to be flushed if the WAL is close to full and the WAL entry is snapset, object_info_t, 
+  collection and pg_log
+  - The premise behind this design is OSD can manage the lateset metadta as a single copy. So,
+    appended entires are not to be read
+- Either best of worst case does not produce severe I/O amplification (it produce I/Os, but I/O rate is constant) 
+  unlike LSM-tree DB (proposed design is similar to LSM-tree which has only level-0)
+  
+
+Detailed Design 
+===============
+
+- Onode lookup
+ - Flat namespace
+  Object data, object meta, omap, and xattrs are all the data related to an object (we call it a head object from ghobject).
+  PoseidonStore adds different suffixes to the head object ID to create PSD_OID to distinguish them
+  (e.g., Assuming head object ID = 12.e4 -> Object data's PSD_OID = 12.e4_data, Object meta's PSD_OID = 12.e4_meta, omap's 
+  PSD_OID = 12.e4_omap_key1, xattr's PSD_OID = 12.e4_xattr_key1).
+  By eliminating suffixes, we can recover their head object ID. Thus, we can know they are all related to the same object.
+  Note that because all objects are sorted by their PSD_OIDs in lexicographic order, xattrs and omap can be iterated 
+  in lexicographic order. For object deletion, all the objects having the same head object ID are searched and deleted.
+
+  A leafnode contains <OID, Block no, Offset, Extent>.
+  With the hash entry, we can find the disk location of the target object.
+  Since the Extent entry in the Block contains <OID, Chunk Number, (Offset, len), (Offset, len), … >, 
+  we can read the object.
+
+ -Lookup
+  As described in the figure, we replaced hash table on Object meta with b+tree to find out items efficiently.
+  This is because we can control the tree based on the policy we define and it can do fast lookup.
+  There is no on-disk onode data structure. To fill the in-memory onode data structure, object meta, xattrs,
+  and omap of the target object can be retrieved using the respective PSD_OIDs.
```

Comment:
How would omap be involved in the onode?

There's a really really good reason not to do it this way: onode lookup (that is, checking for object existence and size) is *by far* the most common read operation -- requiring that the onode be reconsituted from several keys means you need to spend a bunch of cpu cycles doing so.  This alone is one of the largest problems with bluestore as it exists today.  You probably want to arrange for the binary representation of the onode on disk to be contiguous and interpretable directly.  Eventually, we'll want most of the object_info_t to work this way as well, though that will require refactoring elsewhere.